### ISSUE TITLE
[Feature Request] Multiple primals and tangents for tf.autodiff.ForwardAccumulator

### ISSUE THREAD

**Issue by captain-pool** [tensorflow/tensorflow#40074]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): Not Sure

**Describe the feature and the current behavior/state.**
`tf.autodiiff.ForwardAccumulator` only supports Tensors or Variables as primals and tangents. To take JVP of a Neural Network in forward mode, each and every variable is needed to be treated seperately. So, that would mean, number of of iterations needed for the JVP would be equal to the total number of variables. That's completely infeasible.

**Will this change the current api? How?**
Yes, instead of one primal and one tangent in `tf.autodiff.ForwardAccumulator` it would have the capability to take multiple primals and tangents and when .jvp() is called it should return a list of all the JVPs.
**Who will benefit with this feature?**

Anyone who is need forward mode JVPs / HVPs of multiple variables in their code.

**Any Other info.**
Here's a sample of how that should work:
```python3
x = tf.eye(2)
layer = tf.keras.layers.Dense(1)
layer.build([None, 2])
tangents = [tf.eye(*variable.shape) for variable in layer.trainable_variables]
with tf.autodiff.ForwardAccumulator(layer.trainable_variables, tangents) as acc:
  output = layer(x)
  jvps = acc.jvp(output)
```
where `jvps` should be a list of the jvp of the variables and tangents.

CC: @allenlavoie

**Comment by allenlavoie** [comment]
Forward-mode autodiff always computes slices of the output (loss.shape) rather than slices of the input (variables). It's the same shape as finite differences (`(f(x) - f(x + h)) / h`).

To compute a Hessian-vector product, see the ForwardAccumulator docstring: https://github.com/tensorflow/tensorflow/blob/2b96b9c02d81f54dc4959bde2f8b4d92af72679c/tensorflow/python/eager/forwardprop.py#L274-L285

For computing Hessians efficiently you do need to add looping (e.g. from `tf.vectorized_map`): https://github.com/tensorflow/tensorflow/blob/071a3892f6cc2653f38923cd967f62385c3b2f3e/tensorflow/python/eager/forwardprop_test.py#L172-L191

There is a GSoC student this summer working on supporting batches of tangents for each variable, basically pushing `vectorized_map` inside `ForwardAccumulator`. This would allow the forward pass of the model to run eagerly, which you don't get with `vectorized_map` + `ForwardAccumulator` right now (since vectorized_map traces the function). So you'd figure out a batch of tangents that you wanted to compute forwardprop for in advance, and `acc.jvp` would return a corresponding batch of forwardprop runs. Is that something you'd find useful?

**Comment by captain-pool** [comment]
Hey @allenlavoie , this looks great! Thanks for the help. :)

**Comment by IAL32** [comment]
Hi! I would like to contribute to this issue taking over from last year's GSOC work :)
How can I start?

**Comment by bhack** [comment]
@ymodak Can we reopen this? Who could mentor @IAL32?

**Comment by vkarthik095** [comment]
Hi all,
It'll be great if any of you can share a code snippet to generate the jvps mentioned in "_jvps = acc.jvp(output)_". Thanks in advance!
